shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm

```diff
@@ -54,48 +56,70 @@ - (instancetype)init {
 }
 
 - (void)dealloc {
-  [_netService.get() stop];
+  if (_dnsServiceRefInitialized) {
+    DNSServiceRefDeallocate(_dnsServiceRef);
+  }
   blink::DartServiceIsolate::RemoveServerStatusCallback(std::move(_callbackHandle));
   [super dealloc];
 }
 
+- (uint32_t)resolveInterface {
+  // We want to use the loopback interface on the simulator, to force it to be available via
+  // standard mDNS queries.
+#if TARGET_IPHONE_SIMULATOR
+  return if_nametoindex("lo0");
+#else   // TARGET_IPHONE_SIMULATOR
+  return 0;
+#endif  // TARGET_IPHONE_SIMULATOR
+}
+
 - (void)publishServiceProtocolPort:(std::string)uri {
+  if (_dnsServiceRefInitialized) {
+    DNSServiceRefDeallocate(_dnsServiceRef);
+  }
   if (uri.empty()) {
-    [_netService.get() stop];
     return;
   }
   // uri comes in as something like 'http://127.0.0.1:XXXXX/' where XXXXX is the port
   // number.
   NSURL* url =
       [[[NSURL alloc] initWithString:[NSString stringWithUTF8String:uri.c_str()]] autorelease];
 
-  // DNS name has to be a max of 63 bytes.  Prefer to cut off the app name rather than
-  // the device hostName. e.g. 'io.flutter.example@someones-iphone', or
-  // 'ongAppNameBecauseThisCouldHappenAtSomePoint@somelongname-iphone'
-  NSString* serviceName = [NSString
-      stringWithFormat:@"%@@%@",
-                       [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleIdentifier"],
-                       [NSProcessInfo processInfo].hostName];
-  if ([serviceName length] > 63) {
-    serviceName = [serviceName substringFromIndex:[serviceName length] - 63];
-  }
+  NSString* serviceName =
+      [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleIdentifier"];
 
-  _netService.reset([[NSNetService alloc] initWithDomain:@"local."
-                                                    type:@"_dartobservatory._tcp."
-                                                    name:serviceName
-                                                    port:[[url port] intValue]]);
+  DNSServiceFlags flags = kDNSServiceFlagsDefault;
+  uint32_t interfaceIndex = [self resolveInterface];
+  const char* registrationType = "_dartobservatory._tcp";
+  const char* domain = "local.";  // default domain
+  uint16_t port = [[url port] intValue];
 
-  [_netService.get() setDelegate:self];
-  [_netService.get() publish];
-}
+  int err = DNSServiceRegister(&_dnsServiceRef, flags, interfaceIndex, [serviceName UTF8String],
```

Comment:
These methods seem to be [introduced in iOS 9.3+](https://developer.apple.com/documentation/dnssd/1804733-dnsserviceregister) which is higher than our minimum supported version on 8.0. Unfortunately, you will have to have one implementation for the device and one for the simulator (with an overall guard against having any of this is release mode). Sorry.

Reply:
Oof, good catch.  

This will get a little ugly.  And if someone happens to be running an iOS 8.0 simulator (if that's even still possible?), we won't be able to make `flutter attach` work automatically.

Reply:
Not really sure which is uglier - having repeated code, or having guards sprinkled throughout (still with a little repeated code).